### DISCUSSION TITLE
"Error updating branch: update failure" with helm-values and certain dependency names

### DISCUSSION THREAD

**Discussion by mjnagel** [renovatebot/renovate#24880] [Request Help]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Opening issue to follow on to the discussion in https://github.com/renovatebot/renovate/discussions/15306

The basic error I'm seeing is:
```console
DEBUG: Value mismatch(packageFile="chart/values.yaml", branch="renovate/dockerhub")
{
  "manager": "helm-values",
  "expectedValue": "12.10",
  "foundValue": "12.1"
}
WARN: Error updating branch: update failure(branch="renovate/dockerhub")
```

I'm not sure of the root cause here but this error seems to occur with `helm-values` when the image name ends with `12`, which also happens to be the major version. In the main branch the version of this image is `12.9` and the latest available in dockerhub is `12.10`. Somehow it is picking up `12.1` and erroring on it.

Issue was reproduced in [this repo](https://github.com/mjnagel/renovate-testing), after first stumbling on it on the self-hosted side. Additional logs have been included in the below box based on @rarkins reproduction.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Setting current branch to main (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: latest commit (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
       "branchName": "main",
       "latestCommitDate": "2022-04-26T08:15:43-06:00"
DEBUG: getBranchPr(renovate/docker.io-mjnagel-postgresql12-12.x) (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: findPr(renovate/docker.io-mjnagel-postgresql12-12.x, undefined, open) (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: findPr(renovate/docker.io-mjnagel-postgresql12-12.x, undefined, closed) (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: branchExists=false (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: dependencyDashboardCheck=undefined (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: recreateClosed is false (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: findPr(renovate/docker.io-mjnagel-postgresql12-12.x, Update helm values docker.io/mjnagel/postgresql12 to v12.10, !open) (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: prAlreadyExisted=false (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: Checking schedule(at any time, null) (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: No schedule defined (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: Branch needs creating (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: Using reuseExistingBranch: false (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=false (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
DEBUG: Starting search at index 215 (repository=renovate-reproductions/15306, packageFile=chart/values.yaml, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
       "depName": "docker.io/mjnagel/postgresql12"
DEBUG: Found match at index 215 (repository=renovate-reproductions/15306, packageFile=chart/values.yaml, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
       "depName": "docker.io/mjnagel/postgresql12"
DEBUG: Value mismatch (repository=renovate-reproductions/15306, packageFile=chart/values.yaml, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
       "manager": "helm-values",
       "expectedValue": "12.10",
       "foundValue": "12.1"
 WARN: Error updating branch: update failure (repository=renovate-reproductions/15306, branch=renovate/docker.io-mjnagel-postgresql12-12.x)
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by PhilipAbed** [comment]
i debugged, i found this:
```
export function extractPackageFile(content: string): PackageFile | null {
  let parsedContent: Record<string, unknown> | HelmDockerImageDependency;
  try {
    // a parser that allows extracting line numbers would be preferable, with
    // the current approach we need to match anything we find again during the update
    // TODO: fix me (#9610)
    parsedContent = load(content, { json: true }) as any;
```

The `load` function is turning the content from 12.10 to 12.1, 
 a refactor of this function task is here https://github.com/renovatebot/renovate/issues/9610

should we promote the refactor to solve this?

Conclusions:
We are returning type `Any`, and it's causing `JSON` to think that `12.10` is a number instead of a string, which is why the 0 is removed from the end of it.
All this happens in `auto-replace.ts::confirmIfDepUpdated()` functionality.

**Comment by PhilipAbed** [comment]
or do we need to make an update type for each in every PR?
cause this one is helm-values.

**Comment by viceice** [comment]
This is a yaml issue

```yaml
version: 12.10
```

will be converted to a double `12.1`, so it's not a renovate fault. that value should be quoted

**Comment by mjnagel** [comment]
> This is a yaml issue
> 
> ```yaml
> version: 12.10
> ```
> 
> will be converted to a double `12.1`, so it's not a renovate fault. that value should be quoted

@viceice maybe I'm not following...Helm correctly handles this and the 12.10 image will get deployed. I would expect that anything Helm accepts as valid values would also be scraped accordingly by Renovate?

The renovate doc itself shows unquoted tags - https://docs.renovatebot.com/modules/manager/helm-values/#additional-information

**Comment by viceice** [comment]
then helm isn't validating that value correct, as the yaml standard says it's a number

**Comment by mjnagel** [comment]
Putting quotes around the tag does seem to work from a quick test on a self hosted runner. I'm still not sure I quite follow why you see this as a helm/yaml issue...

**Comment by PhilipAbed** [comment]
according to [this](https://stackoverflow.com/questions/19109912/yaml-do-i-need-quotes-for-strings-in-yaml), u can Force a string by adding quotes, but its not necessary, 
adding quotes should be a good workaround though.

i still think our code needs to be able to read it as a string anyway.

**Comment by rarkins** [comment]
If the yaml parser returns a number `12.1` then there's not much we can do other than throwing away yaml parsing

**Comment by MaronHatoum** [comment]
Hi @mjnagel 
can you please see [this](https://github.com/MaronHatoum/renovate-testing/pulls) PR
is this PR the expected result?
because i'm trying to reproduce the issue you reported, this is the [PR](https://github.com/MaronHatoum/renovate-testing/pulls) I got and I can not find the error msg in the log.

**Comment by mjnagel** [comment]
@MaronHatoum if you unquote the tag in the values you will see the issue. I had temporarily updated my reproduction repo, but just reverted the change.

**Comment by MaronHatoum** [comment]
I tried to change the `any` type to an object which has `tag` as a string and the yaml `load` returns `double` anyway.
https://github.com/renovatebot/renovate/blob/0cb19fa99795745aac809b71772f53b1fdf0d4d9/lib/modules/manager/helm-values/extract.ts#L66 
I don't think we can solve this issue easily.
I think the workaround is good enough.

**Comment by rarkins** [comment]
If our YAML parser returns `12.1` when presented with `version: 12.10` then I think we should:
- consult the YAML specs
- file an upstream feature request (or better yet PR) to the parser lib

I can understand if the YAML parser treats `12.1` as a double, but it seems wrong to treat `12.10` as a double and just drop the `0`.

**Comment by MaronHatoum** [comment]
Hi @rarkins 
I found a solution using regex which checks if the version is wrapped by quotes:
- if not it will wrap the version with quotes and then pass the YAML file content to the YAML loader.
- if yes then the YAML file will be passed as is.
I already checked it with unit tests and on a real repository.
PR will be created soon

**Comment by viceice** [comment]
> Hi @rarkins 
> I found a solution using regex which checks if the version is wrapped by quotes:
> - if not it will wrap the version with quotes and then pass the YAML file content to the YAML loader.
> - if yes then the YAML file will be passed as is.
> I already checked it with unit tests and on a real repository.
> PR will be created soon

i don't like that idea, we should check for an upstream fix first.

**Comment by rarkins** [comment]
If this is a problem with values we insert, rather than values we read, then maybe we can solve it ourselves by adding quotes to the newValue. 

We could add quotes for any new version which is a number, or just when they have a trailing zero.

**Comment by nabeelsaabna** [comment]
Pending a decision on how to fix/overcome this.

**Comment by rarkins** [comment]
Decision: if the current version is e.g. `12.9` and new version is `12.10` then we need to ensure that quotation marks are added if they weren't there previously. i.e. if there's a trailing zero like this then we need add quotes. Perhaps also if it's e.g. `13.0`?

**Comment by mjnagel** [comment]
> Decision: if the current version is e.g. `12.9` and new version is `12.10` then we need to ensure that quotation marks are added if they weren't there previously. i.e. if there's a trailing zero like this then we need add quotes. Perhaps also if it's e.g. `13.0`?

I think it would likely be required anytime a version ends with a trailing 0, although I haven't checked this for sure.

**Comment by MaronHatoum** [comment]
thanks for all your inputs
I will update the PR, the fix will be relevant for versions that contain one point and ends with 0.
versions that contain `m.m.p` will be changed automatically to string even if quotes are missing in the YAML file.
